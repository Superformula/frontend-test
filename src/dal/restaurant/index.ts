import { gql } from "@apollo/client";
export { useSearchRestaurantsQuery } from "../../generated/graphql";

/**
 * https://www.yelp.com/developers/graphiql?query=query%20%7B%0A%09search(location%3A%20%22Las%20Vegas%22%2C%20categories%3A%20%22restaurant%22)%20%7B%0A%20%20%20%20business%20%7B%0A%20%20%20%20%20%20name%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D
 *
 */
export const SEARCH_RESTAURANTS = gql`
  query SearchRestaurants(
    $location: String!
    $openNow: Boolean
    $categories: String
    $price: String
    $limit: Int
    $offset: Int
  ) {
    search(
      location: $location
      limit: $limit
      offset: $offset
      categories: $categories
      open_now: $openNow
      price: $price
    ) {
      business {
        ...SearchRestaurantFragment
      }
    }
  }
`;

export const SEARCH_RESTAURANTS_FRAGMET = gql`
  fragment SearchRestaurantFragment on Business {
    name
    is_closed
    rating
    price
    photos
    categories {
      alias
      title
    }
  }
`;
