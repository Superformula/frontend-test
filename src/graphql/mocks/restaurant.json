{
  "business": {
    "__typename": "Business",
    "name": "Mon Ami Gabi",
    "rating": 4,
    "categories": [
      { "__typename": "Category", "title": "French" },
      { "__typename": "Category", "title": "Steakhouses" },
      { "__typename": "Category", "title": "Breakfast & Brunch" }
    ],
    "price": "$$",
    "is_closed": false,
    "photos": [
      "https://s3-media1.fl.yelpcdn.com/bphoto/tEcdu80k_iS4OIqEVSDkZw/o.jpg",
      "https://s3-media1.fl.yelpcdn.com/bphoto/tEcdu80k_iS4OIqEVSDkZw/o.jpg",
      "https://s3-media1.fl.yelpcdn.com/bphoto/tEcdu80k_iS4OIqEVSDkZw/o.jpg"
    ],
    "location": {
      "__typename": "Location",
      "formatted_address": "3655 Las Vegas Blvd S\nLas Vegas, NV 89109"
    },
    "review_count": 9737,
    "reviews": [
      {
        "__typename": "Review",
        "rating": 5,
        "text": "Our server for sure wanted us to leave...  but why would we?  This is one of the most romantic patios in Las Vegas.  The view of the Bellagio fountains is...",
        "time_created": "2020-11-13 23:56:08",
        "user": {
          "__typename": "User",
          "name": "Crystal L.",
          "image_url": "https://s3-media3.fl.yelpcdn.com/photo/gjz8X6tqE3e4praK4HfCiA/o.jpg"
        }
      },
      {
        "__typename": "Review",
        "rating": 5,
        "text": "I'm local and shamelessly visit Mon Ami at least once a month. I love it. Mon Ami does its best to be efficient, and at full seating capacity, they do a...",
        "time_created": "2020-11-28 10:16:21",
        "user": {
          "__typename": "User",
          "name": "Jocelyn C.",
          "image_url": "https://s3-media4.fl.yelpcdn.com/photo/bJj1XlSB5AfImruJ_DvXHQ/o.jpg"
        }
      },
      {
        "__typename": "Review",
        "rating": 5,
        "text": "My first time here and it is absolutely worth the hype! This restaurant is located in the Paris hotel and is absolutely gorgeous! Super romantic and a great...",
        "time_created": "2020-12-06 17:10:05",
        "user": {
          "__typename": "User",
          "name": "Athina L.",
          "image_url": "https://s3-media4.fl.yelpcdn.com/photo/O4h96A0teB2lkprPr-xccg/o.jpg"
        }
      }
    ]
  }
}
