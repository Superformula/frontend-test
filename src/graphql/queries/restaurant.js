import { gql } from '@apollo/client';

export const restaurantQuery = gql`
  query GetRestaurantDetails($id: String, $offset: Int = 0, $limit: Int = 6) {
    business(id: $id) {
      name
      rating
      categories {
        title
      }
      price
      is_closed
      photos
      location {
        formatted_address
      }
      review_count
      reviews(offset: $offset, limit: $limit) {
        rating
        text
        time_created
        user {
          name
          image_url
        }
      }
    }
  }
`;
