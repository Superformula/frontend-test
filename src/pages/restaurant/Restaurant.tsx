import React from "react";
import styled from "@emotion/styled";
import { useQuery } from "@apollo/react-hooks";
import { useParams } from "react-router";
import { query } from "./graphql";
import { css } from "emotion";
import ReactMapGL, { Marker } from "react-map-gl";

import { Hero } from "../../components/Hero";
import { Comment } from "../../components/Comment";
import { RestaurantStarRating } from "../../components/RestaurantStarRating";
import MarkerIcon from "../../components/icons/MarkerIcon";
import DotIcon from "../../components/icons/DotIcon";

function Restaurant() {
  let { id } = useParams();
  const { loading, data, error, refetch } = useQuery(query, {
    variables: { id }
  });

  if (loading) return <div>Loading...</div>;
  if (error) return <div>{error.message}</div>;

  const { business } = data;

  return (
    <Main>
      <Summary className="summary">
        <Hero description="" title={business.name}></Hero>
        <RestaurantStarRating rating={business.rating}></RestaurantStarRating>
        <div
          className={css`
            display: flex;
            justify-content: space-between;
            color: #666666;
            font-size: 22px;
            line-height: 32px;
          `}>
          <span>
            {business.categories[0].title} â€¢ {business.price}
          </span>
          <span
            className={css`
              display: flex;
              align-items: center;
            `}>
            <DotIcon
              className={css`
                height: 1em;
              `}
              filled={business.hours[0].is_open_now}></DotIcon>
            <span
              className={css`
                margin-left: 0.5em;
              `}>
              {business.hours[0].is_open_now ? "Open Now" : "Closed"}
            </span>
          </span>
        </div>
      </Summary>
      <Detail className="detail">
        <div
          className={css`
            display: flex;
          `}>
          <ReactMapGL
            width={400}
            height={200}
            latitude={business.coordinates.latitude}
            longitude={business.coordinates.longitude}
            zoom={11}
            mapboxApiAccessToken={process.env.MAPBOX_ACCESS_TOKEN}
            onViewportChange={viewport => {
              const { width, height, latitude, longitude, zoom } = viewport;
              // Optionally call `setState` and use the state to update the map.
            }}>
            <Marker
              latitude={business.coordinates.latitude}
              longitude={business.coordinates.longitude}
              offsetTop={-40}>
              <MarkerIcon></MarkerIcon>
              {/* <div
                className={css`
                  background-color: white;
                  border-radius: 4px;
                  max-width: 8rem;
                `}>
                <p>{business.name}</p>
              </div> */}
            </Marker>
          </ReactMapGL>
          {business.photos.map(photo => (
            <img
              key={photo}
              className={css`
                height: 200px;
                margin-left: 20px;
              `}
              src={photo}
            />
          ))}
        </div>
      </Detail>
      <CommentSummary className="comments">
        <div
          className={css`
            font-size: 22px;
            line-height: 32px;
            letter-spacing: 0.916667px;
            color: #666666;
          `}>
          {business.review_count} Reviews
        </div>
        {business.reviews.map(review => (
          <Comment review={review}></Comment>
        ))}
      </CommentSummary>
    </Main>
  );
}

export default Restaurant;

const Main = styled("div")`
  display: grid;
  width: 100vw;
  height: 100vh;
  grid-template-rows: auto auto 1fr;
  grid-template-areas:
    "summary"
    "detail"
    "comments";

  .summary {
    grid-area: summary;
  }
  .detail {
    grid-area: detail;
  }
  .comments {
    grid-area: comments;
  }
`;

const Summary = styled("div")`
  padding: 2em 40px;
`;
const Detail = styled("div")`
  padding: 2em;
  border-top: solid #e6e6e6 1px;
  border-bottom: solid #e6e6e6 1px;
`;
const CommentSummary = styled("div")`
  margin: 20px 40px;
  .comment-item {
    border-bottom: solid #e6e6e6 1px;
    padding-bottom: 2em;
  }
`;
