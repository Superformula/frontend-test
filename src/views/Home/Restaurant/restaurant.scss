.restaurant {
  display: flex;
  flex-direction: column;
  min-width: 0;
  min-height: 0;
  
  .photo {
    width: 100%;
  }
}