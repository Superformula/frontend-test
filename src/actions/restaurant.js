import api from 'api';
import { sleep } from 'helpers/sleep';
import ENV from 'helpers/env';
import { buildQueryObject } from 'reducers/helpers';
import { getFilters } from 'selectors';

const { YELP_API_ROOT } = ENV;

export const FETCH_RESTAURANTS_START = 'FETCH_RESTAURANTS_START';
export const FETCH_RESTAURANTS_SUCCESS = 'FETCH_RESTAURANTS_SUCCESS';
export const FETCH_RESTAURANTS_ERROR = 'FETCH_RESTAURANTS_ERROR';

export const FETCH_RESTAURANT_START = 'FETCH_RESTAURANT_START';
export const FETCH_RESTAURANT_SUCCESS = 'FETCH_RESTAURANT_SUCCESS';
export const FETCH_RESTAURANT_ERROR = 'FETCH_RESTAURANT_ERROR';

export const FETCH_RESTAURANT_REVIEWS_START = 'FETCH_RESTAURANT_REVIEWS_START';
export const FETCH_RESTAURANT_REVIEWS_SUCCESS =  'FETCH_RESTAURANT_REVIEWS_SUCCESS'; // eslint-disable-line
export const FETCH_RESTAURANT_REVIEWS_ERROR = 'FETCH_RESTAURANT_REVIEWS_ERROR';

export const TOGGLE_FILTER = 'TOGGLE_FILTER';
export const RESET_FILTERS = 'RESET_FILTERS';

export function getRestaurants() {
  return async (dispatch, getState) => {
    dispatch({
      type: FETCH_RESTAURANTS_START,
    });

    await sleep(500);

    const filters = getFilters(getState());

    try {
      const { data } = await api.get(`${YELP_API_ROOT}/businesses/search`, {
        params: buildQueryObject(filters),
      });
      return dispatch({
        type: FETCH_RESTAURANTS_SUCCESS,
        data,
      });
    } catch (error) {
      return dispatch({
        type: FETCH_RESTAURANTS_ERROR,
        error,
      });
    }
  };
}

export function getRestaurant(id) {
  return async (dispatch) => {
    dispatch({
      type: FETCH_RESTAURANT_START,
    });

    await sleep(500);

    try {
      const { data } = await api.get(`${YELP_API_ROOT}/businesses/${id}`);
      return dispatch({
        type: FETCH_RESTAURANT_SUCCESS,
        data,
      });
    } catch (error) {
      return dispatch({
        type: FETCH_RESTAURANT_ERROR,
        error,
      });
    }
  };
}

export function getRestaurantReviews(id) {
  return async (dispatch) => {
    dispatch({
      type: FETCH_RESTAURANT_REVIEWS_START,
    });

    await sleep(500);

    try {
      const { data } = await api.get(
        `${YELP_API_ROOT}/businesses/${id}/reviews`,
      );
      return dispatch({
        type: FETCH_RESTAURANT_REVIEWS_SUCCESS,
        data,
      });
    } catch (error) {
      return dispatch({
        type: FETCH_RESTAURANT_REVIEWS_ERROR,
        error,
      });
    }
  };
}

export function toggleFilter(filter, value) {
  return (dispatch) => {
    dispatch({
      type: TOGGLE_FILTER,
      filter,
      value,
    });

    dispatch(getRestaurants());
  };
}

export function resetFilters() {
  return async (dispatch) => {
    await dispatch({
      type: RESET_FILTERS,
    });

    dispatch(getRestaurants());
  };
}
