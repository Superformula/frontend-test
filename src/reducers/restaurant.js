import * as ACTIONS from 'actions/restaurant';

import Helpers from 'reducers/helpers';

const initialState = {
  isFetching: false,
  restaurants: [],
  restaurant: {},
  reviews: [],
  filters: {
    location: 'Las Vegas', // hardcoding here; TODO: add searchable field
  },
};

const actionsMap = {
  [ACTIONS.FETCH_RESTAURANTS_START]: (state, { data }) => {
    const nextState = Object.assign({}, state, {
      isFetching: true,
    });

    return nextState;
  },
  [ACTIONS.FETCH_RESTAURANT_START]: (state, { data }) => {
    const nextState = Object.assign({}, state, {
      isFetching: true,
    });

    return nextState;
  },
  [ACTIONS.FETCH_RESTAURANT_REVIEWS_START]: (state, { data }) => {
    const nextState = Object.assign({}, state, {
      isFetching: true, // TODO: consider using a separate reducer or key for reviews loading
    });

    return nextState;
  },
  [ACTIONS.FETCH_RESTAURANT_START]: (state, { data }) => {
    const nextState = Object.assign({}, state, {
      isFetching: true,
    });

    return nextState;
  },
  [ACTIONS.FETCH_RESTAURANTS_SUCCESS]: (state, { data }) => {
    const nextState = Object.assign({}, state, {
      restaurants: data.businesses,
      isFetching: false,
    });

    return nextState;
  },
  [ACTIONS.FETCH_RESTAURANT_SUCCESS]: (state, { data }) => {
    const nextState = Object.assign({}, state, {
      restaurant: data,
      isFetching: false,
    });

    return nextState;
  },
  [ACTIONS.FETCH_RESTAURANT_REVIEWS_SUCCESS]: (state, { data }) => {
    const nextState = Object.assign({}, state, {
      reviews: data.reviews,
      isFetching: false,
    });

    return nextState;
  },
  [ACTIONS.TOGGLE_FILTER]: (state, { filter, value }) => {
    const nextState = Object.assign({}, state, {
      filters: Helpers.toggleFilter(state, filter, value),
    });

    return nextState;
  },
  [ACTIONS.RESET_FILTERS]: (state) => {
    const nextState = Object.assign({}, state, {
      filters: { ...initialState.filters },
    });

    return nextState;
  },
};

export default function reducer(state = initialState, action = {}) {
  const fn = actionsMap[action.type];
  return fn ? fn(state, action) : state;
}
