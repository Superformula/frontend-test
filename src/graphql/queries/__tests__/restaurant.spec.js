import { restaurantQuery } from '../restaurant';

describe('restaurantQuery', () => {
  it('should build correct restaurant query', () => {
    expect(restaurantQuery).toMatchSnapshot();
  });
});
