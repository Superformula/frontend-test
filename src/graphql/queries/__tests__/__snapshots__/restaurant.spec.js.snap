// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`restaurantQuery should build correct restaurant query 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "name": Object {
        "kind": "Name",
        "value": "GetRestaurantDetails",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "id",
                },
                "value": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                },
              },
            ],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "business",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "name",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "rating",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "categories",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "title",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "price",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "is_closed",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "photos",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "location",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "formatted_address",
                        },
                        "selectionSet": undefined,
                      },
                    ],
                  },
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "review_count",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "offset",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "offset",
                        },
                      },
                    },
                    Object {
                      "kind": "Argument",
                      "name": Object {
                        "kind": "Name",
                        "value": "limit",
                      },
                      "value": Object {
                        "kind": "Variable",
                        "name": Object {
                          "kind": "Name",
                          "value": "limit",
                        },
                      },
                    },
                  ],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "reviews",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "rating",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "text",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "time_created",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "user",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "image_url",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [
        Object {
          "defaultValue": undefined,
          "directives": Array [],
          "kind": "VariableDefinition",
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
          "variable": Object {
            "kind": "Variable",
            "name": Object {
              "kind": "Name",
              "value": "id",
            },
          },
        },
        Object {
          "defaultValue": Object {
            "kind": "IntValue",
            "value": "0",
          },
          "directives": Array [],
          "kind": "VariableDefinition",
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Int",
            },
          },
          "variable": Object {
            "kind": "Variable",
            "name": Object {
              "kind": "Name",
              "value": "offset",
            },
          },
        },
        Object {
          "defaultValue": Object {
            "kind": "IntValue",
            "value": "6",
          },
          "directives": Array [],
          "kind": "VariableDefinition",
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Int",
            },
          },
          "variable": Object {
            "kind": "Variable",
            "name": Object {
              "kind": "Name",
              "value": "limit",
            },
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 455,
    "start": 0,
  },
}
`;
